File: ETour-DotNet/Common/CategoryScope.cs
using ETour.Api.Models;

namespace ETour.Api.Common;

public static class CategoryScope
{
    private const int MaxDepth = 20;

    private static readonly Dictionary<string, HashSet<string>> Themes = new()
    {
        ["SPT"] = new HashSet<string> { "CKD", "ECD", "SLC" },
        ["CKD"] = new HashSet<string> { "ECD", "SLC" },
        ["ECD"] = new HashSet<string> { "CKD", "SLC" },
        ["SLC"] = new HashSet<string> { "CKD", "ECD" }
    };

    public static HashSet<int> Resolve(int? categoryId, IEnumerable<Category> allCategories)
    {
        var scope = new HashSet<int>();
        if (categoryId is null)
        {
            return scope;
        }

        var childrenOf = new Dictionary<int, List<int>>();
        var codeOf = new Dictionary<int, string>();
        var idsByCode = new Dictionary<string, List<int>>();

        foreach (var c in allCategories)
        {
            codeOf[c.CategoryId] = c.CatCode;

            if (c.CatCode is not null)
            {
                if (!idsByCode.TryGetValue(c.CatCode, out var byCode))
                {
                    byCode = new List<int>();
                    idsByCode[c.CatCode] = byCode;
                }
                byCode.Add(c.CategoryId);
            }

            if (c.ParentId is int parentId)
            {
                if (!childrenOf.TryGetValue(parentId, out var kids))
                {
                    kids = new List<int>();
                    childrenOf[parentId] = kids;
                }
                kids.Add(c.CategoryId);
            }
        }

        AddWithDescendants(categoryId.Value, childrenOf, scope);

        if (codeOf.TryGetValue(categoryId.Value, out var code) && code is not null
            && Themes.TryGetValue(code, out var members))
        {
            foreach (var member in members)
            {
                if (!idsByCode.TryGetValue(member, out var ids)) continue;
                foreach (var id in ids)
                {
                    AddWithDescendants(id, childrenOf, scope);
                }
            }
        }

        return scope;
    }

    private static void AddWithDescendants(int rootId,
                                           Dictionary<int, List<int>> childrenOf,
                                           HashSet<int> into)
    {
        var queue = new Queue<int>();
        queue.Enqueue(rootId);

        var guard = 0;
        while (queue.Count > 0 && guard++ < MaxDepth * 100)
        {
            var current = queue.Dequeue();
            if (!into.Add(current))
            {
                continue;
            }
            if (childrenOf.TryGetValue(current, out var kids))
            {
                foreach (var k in kids) queue.Enqueue(k);
            }
        }
    }
}

File: E-Tour/src/main/java/com/etour/common/CategoryScope.java
package com.etour.common;

import java.util.ArrayDeque;
import java.util.Collection;
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.etour.entities.Category;

public final class CategoryScope {

    private static final int MAX_DEPTH = 20;

    private static final Map<String, Set<String>> THEMES = Map.of(
            "SPT", Set.of("CKD", "ECD", "SLC"),
            "CKD", Set.of("ECD", "SLC"),
            "ECD", Set.of("CKD", "SLC"),
            "SLC", Set.of("CKD", "ECD")
    );

    private CategoryScope() {
    }

    public static Set<Integer> resolve(Integer categoryId, Collection<Category> allCategories) {

        Set<Integer> scope = new LinkedHashSet<>();
        if (categoryId == null) {
            return scope;
        }

        Map<Integer, List<Integer>> childrenOf = new HashMap<>();
        Map<Integer, String> codeOf = new HashMap<>();
        Map<String, List<Integer>> idsByCode = new HashMap<>();

        for (Category c : allCategories) {
            if (c.getCategoryId() == null) {
                continue;
            }
            codeOf.put(c.getCategoryId(), c.getCatCode());

            if (c.getCatCode() != null) {
                idsByCode.computeIfAbsent(c.getCatCode(), k -> new java.util.ArrayList<>())
                         .add(c.getCategoryId());
            }
            if (c.getParent() != null && c.getParent().getCategoryId() != null) {
                childrenOf.computeIfAbsent(c.getParent().getCategoryId(),
                                           k -> new java.util.ArrayList<>())
                          .add(c.getCategoryId());
            }
        }

        addWithDescendants(categoryId, childrenOf, scope);

        String code = codeOf.get(categoryId);
        if (code != null) {
            for (String member : THEMES.getOrDefault(code, Set.of())) {
                for (Integer id : idsByCode.getOrDefault(member, List.of())) {
                    addWithDescendants(id, childrenOf, scope);
                }
            }
        }

        return scope;
    }

    private static void addWithDescendants(Integer rootId,
                                           Map<Integer, List<Integer>> childrenOf,
                                           Set<Integer> into) {

        Deque<Integer> queue = new ArrayDeque<>();
        queue.add(rootId);

        int guard = 0;
        while (!queue.isEmpty() && guard++ < MAX_DEPTH * 100) {
            Integer current = queue.poll();
            if (current == null || !into.add(current)) {
                continue;
            }
            queue.addAll(childrenOf.getOrDefault(current, List.of()));
        }
    }
}
